src/crimson/os/seastore/journal/circular_bounded_journal.h

```diff
@@ -0,0 +1,286 @@
+    uint32_t error;      // reserved
+
+    rbm_abs_addr start_offset;      // start offset of CBJournal
+    rbm_abs_addr last_committed_record_base;
+    rbm_abs_addr written_to;
+    rbm_abs_addr applied_to;
+
+    uint64_t flag;       // represent features (reserved)
+    uint8_t csum_type;   // type of checksum algoritghm used in cbj_header_t
+    uint64_t csum;       // checksum of entire cbj_header_t
+    uint32_t cur_segment_seq;
+
+    rbm_abs_addr start; // start address of the device
+    rbm_abs_addr end;   // start address of the device
+    device_id_t device_id;
+
+    DENC(cbj_header_t, v, p) {
+      DENC_START(1, 1, p);
+      denc(v.magic, p);
+      denc(v.uuid, p);
+      denc(v.block_size, p);
+      denc(v.size, p);
+      denc(v.used_size, p);
+      denc(v.error, p);
+
+      denc(v.start_offset, p);
+
+      denc(v.last_committed_record_base, p);
+      denc(v.written_to, p);
+      denc(v.applied_to, p);
+
+      denc(v.flag, p);
+      denc(v.csum_type, p);
+      denc(v.csum, p);
+      denc(v.cur_segment_seq, p);
+      denc(v.start, p);
+      denc(v.end, p);
+      denc(v.device_id, p);
+
+      DENC_FINISH(p);
+    }
+  };
+
+  /**
+   *
+   * Write position for CBJournal
+   *
+   * | written to rbm |    written length to CBJournal    | new write |
+   * ----------------->----------------------------------->------------>
+   *                  ^      	            ^             ^
+   *            applied_to   last_committed_record_base   written_to
+   *
+   */
+
+  size_t get_used_size() const {
+    return header.used_size;
+  }
+  void set_used_size(size_t size) {
+    header.used_size = size;
+  }
+  size_t get_total_size() const {
+    return header.size;
+  }
+  rbm_abs_addr get_start_addr() const {
+    return header.start_offset;
+  }
+  size_t get_available_size() const {
+    return get_total_size() - get_used_size();
+  }
+
+  void update_applied_to(rbm_abs_addr addr, uint32_t len) {
+    rbm_abs_addr new_applied_to = addr;
+    set_used_size(
+      get_last_committed_record_base() >= new_applied_to ?
+      get_written_to() - (new_applied_to + len) :
+      get_written_to() + get_total_size() - (new_applied_to + len));
+    set_applied_to(new_applied_to + len);
+  }
+
+  write_ertr::future<> sync_super();
+
+  read_record_ret return_record(record_group_header_t& header, bufferlist bl);
+
+  void set_write_pipeline(WritePipeline *_write_pipeline) final {
+    write_pipeline = _write_pipeline;
+  }
+
+  rbm_abs_addr get_written_to() const {
+    return header.written_to;
+  }
+  void set_written_to(rbm_abs_addr addr) {
+    header.written_to = addr;
+  }
+  rbm_abs_addr get_last_committed_record_base() const {
+    return header.last_committed_record_base;
+  }
+  void set_last_committed_record_base(rbm_abs_addr addr) {
+    header.last_committed_record_base = addr;
+  }
+  rbm_abs_addr get_applied_to() const {
+    return header.applied_to;
+  }
+  void set_applied_to(rbm_abs_addr addr) {
```

Comment:
This only gets called from update_applied_to, which does not seem to have any callers.